DMA Support for the LPC55S3X Series

EmilioCBen opened the pull request:
Added DMA Support for the LPC55S3X Series.

EmilioCBen committed:
boards: arm: lpcxpresso55s36: Added DMA Support for LPC55S36

Added DMA Support for the LPC55S36 Board as well as
a creation for space on the HEAP when DMA Blocks are
required.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen committed:
tests: drivers: dma: chan_blen_transfer: Added Overlay file.

Added an overlay file for the LPC55S36 to demonstrate
the DMA Support with the chan_blen_transfer test.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

decsny commented on boards/arm/lpcxpresso55s36/Kconfig.defconfig:
```diff
@@ -9,3 +9,13 @@ config BOARD
+# channels that use multiple blocks for a DMA transfer.
+# Adjust HEAP_MEM_POOL_SIZE in case you need more memory.
+config HEAP_MEM_POOL_SIZE
+	default 4096
```
this seems strange because it might make the heap pool size smaller than otherwise needed for something else. Can we do: `if DMA_MCUX_LPC && HEAP_MEM_POOL_SIZE < 4096`

It looks like the kconfig language supports this: https://www.kernel.org/doc/html/latest/kbuild/kconfig-language.html#menu-dependencies

mmahadevan108 replied:
> What is the heap needed for here?

There are channel specific data structures allocated in the DMA driver. It is allocated based on the number of DMA channels needed by the application via device tree.

mmahadevan108 replied:
This is no longer needed, please delete.

EmilioCBen replied:
Thank you, I have removed this.

decsny commented on tests/drivers/dma/chan_blen_transfer/boards/lpcxpresso55s36.overlay:
```diff
@@ -0,0 +1 @@
+test_dma0: &dma0 { };
```
should this board be added to the test yaml?

EmilioCBen replied:
the testcase.yaml is setup for all dma supported devices on the chan blen test

decsny replied:
okay

decsny commented on boards/arm/lpcxpresso55s36/lpcxpresso55s36.dts:
```diff
@@ -162,3 +162,8 @@
+&dma0 {
+	status = "okay";
+	dma-channels = <52>;
+};
```
isn't there also a dma1?

EmilioCBen replied:
DMA1 will be included in the next push to this branch.

teburd reviewed (commented):
What is the heap needed for here?

mmahadevan108 commented:
@EmilioCBen, was this tested with the dma tests under `tests/drivers/dma/loop_transfer/` and `tests/drivers/dma/scatter_gather`

EmilioCBen committed:
dts: arm: nxp: lpc55s3x: Added DMA Nodes in dts files.

Added dts nodes for DMA support on LPC55S3X devices.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

mmahadevan108 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -740,40 +740,40 @@ static const struct dma_mcux_lpc_config dma_##n##_config = {		\
+			[DT_INST_PROP_OR(n, nxp_dma_num_of_otrigs, 0)];		\
+										\
+	static int8_t								\
+		dma_##n##_channel_index_arr[TOTAL_DMA_CHANNELS(n)] = {0};	\
```
I do not understand this change. You are indirectly passing `n`

danieldegrasse replied:
@EmilioCBen can you take another look at this change? It seems to be causing a build failure in the UART async API test.

decsny replied:
@danieldegrasse that build failure was because the uart node needed its dmas phandle property, which I think emilio has fixed

but I also do not like this SDK-based macro, I see no reason not to use DT_INST_PROP_OR to get the number of channels like the array two lines above uses for the same purpose

EmilioCBen replied:
Something is definitely off, the phandle did not fix the build issues according to twister but locally I don't not have build errors, but the tests do not pass, despite the correct number of channels being passed.

EmilioCBen replied:
I also do not like the SDK Macro here and we are now using several ideas to do the same thing in this driver macro define, and there may be a bigger picture to the SDK macro that I am not aware of, but we should discuss a single way to set the size of these arrays.

EmilioCBen replied:
I tested the changes on the PR, no build error but again the test suite fails.

EmilioCBen replied:
Build errors resolved

mmahadevan108 reviewed (dismissed):
Please add overlay and test with the DMA tests `loop_transfer` and `scatter_gather`

EmilioCBen committed:
drivers: dma: dma_mcux_lpx: Added parameter in macro for 55S36

Added a parameter inside the Channel Number macro since
the LPC55S36 expects an address rather than a static number.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen commented:
> @EmilioCBen, was this tested with the dma tests under `tests/drivers/dma/loop_transfer/` and `tests/drivers/dma/scatter_gather`

This has not yet been tested with loop or scatter, I can work on this

EmilioCBen committed:
tests: drivers: dma: loop_transfer: Added Overlay file.

Added an overlay file for the LPC55S36 to demonstrate
the DMA Support with the loop_transfer test.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen committed:
tests: drivers: dma: scatter_gather: Added Overlay file.

Added an overlay file for the LPC55S36 to demonstrate
the DMA Support with the scatter_gather test.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen commented:
> Please add overlay and test with the DMA tests `loop_transfer` and `scatter_gather`

I have made an overlay file for both of these tests now and they both pass, thank you.

decsny commented on boards/arm/lpcxpresso55s36/lpcxpresso55s36.dts:
```diff
@@ -168,3 +170,15 @@ zephyr_udc0: &usbfs {
 };
+
+&dma0 {
+	dma-channels = <52>;
```
dma-channels should be set at SOC level (see activity in #59767, this is a bug being currently fixed, don't add to it)

EmilioCBen replied:
I have moved these to the SOC level

decsny commented on boards/arm/lpcxpresso55s36/lpcxpresso55s36.dts:
```diff
@@ -168,3 +170,15 @@ zephyr_udc0: &usbfs {
+
+&dma0 {
+	dma-channels = <52>;
+	nxp,dma-num-of-otrigs = <4>;
```
number of otrigs should be set at SOC level

EmilioCBen replied:
These are now in the SOC levels

decsny commented:
the twister failures are because you have not provided peripherals with DMA request numbers with the dmas phandles property in devicetree

mmahadevan108 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -753,7 +753,6 @@ static const struct dma_mcux_lpc_config dma_##n##_config = {		\
 			    &dma_mcux_lpc_api);				\
 									\
 	DMA_MCUX_LPC_CONFIG_FUNC(n)					\
-									\
```
Please remove this change

EmilioCBen commented on boards/arm/lpcxpresso55s36/lpcxpresso55s36.dts:
```diff
@@ -120,6 +120,8 @@
 	pinctrl-0 = <&pinmux_flexcomm0_usart>;
 	pinctrl-names = "default";
+	dmas = <&dma0 4>, <&dma0 5>;
+	dma-names = "rx", "tx";
```
Updated to reflect this comment @decsny please have another look at this.

mmahadevan108 commented:
@EmilioCBen Please resolve the conflicts,

EmilioCBen commented:
> @EmilioCBen Please resolve the conflicts,

Got I have rebased and fixed the conflicts

danieldegrasse commented on tests/drivers/dma/loop_transfer/boards/lpcxpresso55s36.overlay:
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright C 2023 NXP
```
Sorry, missed this- please fix the copyright string here

EmilioCBen replied:
Apologies, this is fixed now

danieldegrasse reviewed (changes requested):
Just requesting changes because the copyright should be updated before this merges, please feel free to ping me when updated @EmilioCBen
